### PR TITLE
549 savant pytorch integration

### LINKED ISSUE #549: Savant PyTorch Integration
Demonstrate the use of PyTorch GPU inference in Savant PyFunc.

### PR DESCRIPTION
**Pull request by dorgun**

### CODE REVIEW

File: docs/source/advanced_topics/12_torch_hub.rst [hunk 1]
```diff
@@ -0,0 +1,11 @@
+TorchHub
+--------
+
+`TorchHub <https://pytorch.org/hub/>`__ is a tool for publishing and using pre-trained models. This approach is not applicable for production use, but using TorchHub and Savant you can test models on real video streams or videos.
+
+The `official documentation <https://pytorch.org/docs/stable/hub.html#loading-models-from-hub>`__  describes how to load the model and perform the inference you can.
```
Comment (abramov-oleg) [inline docs/source/advanced_topics/12_torch_hub.rst]:
```suggestion
The `official documentation <https://pytorch.org/docs/stable/hub.html#loading-models-from-hub>`__  describes how to load the model and perform the inference.
```

File: docs/source/advanced_topics/12_torch_hub.rst [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+Using Savant you can easily make an inference on a frame or for an object of interest (`example <https://github.com/insight-platform/Savant/tree/develop/samples/panoptic_driving_perception>`__. Also using the conversion function (:ref:`advanced_topics/11_memory_representation_function:Conversions Between GPU Memory Formats`), you can get the image as a pytorch tensor on GPU and perform the necessary preprocessing. This will avoid copying data to the host memory and back to the GPU.
```
Comment (abramov-oleg) [inline docs/source/advanced_topics/12_torch_hub.rst]:
```suggestion
With Savant it is easy to run Pytorch inference on a frame or for an object of interest. Look at the `panoptic_driving_perception <https://github.com/insight-platform/Savant/tree/develop/samples/panoptic_driving_perception>`__ sample to learn how. Note how OpenCV GPUMat <-> Pytorch GPU tensor conversion function (:ref:`advanced_topics/11_memory_representation_function:Conversions Between GPU Memory Formats`) makes it possible to get the image as a Pytorch GPU tensor and perform the necessary preprocessing exactly as in a strictly Pytorch pipeline, while avoiding copying data to the host memory and back to the GPU.
```

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import cv2
+import torch
+import torchvision
+import torchvision.transforms as transforms
+
+from savant.deepstream.meta.frame import NvDsFrameMeta
+from savant.deepstream.opencv_utils import alpha_comp, nvds_to_gpu_mat
+from savant.deepstream.pyfunc import NvDsPyFuncPlugin
+from savant.gstreamer import Gst
+from savant.utils.artist import Artist
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: Import is not used.

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+class PyTorchInfer(NvDsPyFuncPlugin):
+    """Custom frame processor."""
+
+    def __init__(
+        self, conf_threshold, iou_threshold, road_mask_color, line_mask_color, **kwargs
+    ):
+        super().__init__(**kwargs)
+        import sys
+
+        print(sys.path)
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: Debug print? Better to remove, I think.

File: docs/source/advanced_topics/12_torch_hub.rst [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+.. note::
+    Be careful when installing dependencies necessary for the model to work, because explicitly or implicitly a version of OpenCV without CUDA support may be installed in the Savant docker container. This will cause errors and make it impossible to run the code.
```
Comment (abramov-oleg) [inline docs/source/advanced_topics/12_torch_hub.rst]:
```suggestion
.. note::

    Be careful when installing dependencies and additional packages. Savant docker container comes with OpenCV with CUDA support preinstalled, and using other OpenCV versions may cause errors and make it impossible to run the code.
```

File: samples/panoptic_driving_perception/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+# open 'rtsp://127.0.0.1:554/stream' in your player
+# or visit 'http://127.0.0.1:888/stream/' (LL-HLS)
+
+# Ctrl+C to stop running the compose bundle
+
+# to get back to project root
+cd ../..
+```
+
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/README.md:49]: One line too many at the end of the file, let's remove.

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        self.model = torch.hub.load('hustvl/yolop', 'yolop', pretrained=True)
+        self.model.cuda()
+        self.model.eval()
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: Currently, the model is loaded in fp32 mode.

Seems like we usually use fp16 and it wouldn't be too difficult to do fp16 inference here as well.

Up to you.

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    def process_frame(self, buffer: Gst.Buffer, frame_meta: NvDsFrameMeta):
+        """Process frame.
+
+        :param buffer: GStreamer buffer.
+        :param frame_meta: Processed frame metadata.
+        """
+        stream = self.get_cuda_stream(frame_meta)
+        with nvds_to_gpu_mat(buffer, frame_meta.frame_meta) as frame_mat:
+            with torch.no_grad():
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: The modern replacement for the `no_grad` mode in Pytorch is [inference_mode](https://github.com/ailzhang/rfcs/blob/rfc0011/RFC-0011-InferenceMode.md) / [doc](https://pytorch.org/docs/stable/generated/torch.inference_mode.html)

Suggest using `inference_mode`

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                input_image = cv2.cuda.resize(frame_mat, (640, 480), stream=stream)
+
+                input_tensor = opencv_gpu_mat_as_pytorch_tensor(input_image).permute(
+                    2, 0, 1
+                )
+                input_tensor = input_tensor[:3, :, :].float() / 255
+                input_tensor = self.normalize(input_tensor).unsqueeze(0)
+                det_out, da_seg_out, ll_seg_out = self.model(input_tensor)
+                da_seg_out = da_seg_out.detach()
+                ll_seg_out = ll_seg_out.detach()
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: Should be a no-op under `no_grad`, suggest removing.

File: samples/panoptic_driving_perception/pytorch_infer.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+                element_name='yolop',
+                label='car',
+                bbox=bbox,
+                confidence=float(obj_meta_tensor[4]),
+            )
+            frame_meta.add_obj_meta(obj_meta)
+
+    def on_source_eos(self, source_id: str):
+        """On source EOS event callback."""
+        self.logger.debug('Got GST_NVEVENT_STREAM_EOS for source %s.', source_id)
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/pytorch_infer.py]: Doesn't seem useful, suggest removing.

File: samples/panoptic_driving_perception/docker/Dockerfile.l4t [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+RUN pip3 install --no-cache-dir --verbose /opt/torchvision*.whl && rm /opt/torchvision*.whl
+
+ENV TORCH_HOME=/cache/models/torch_hub
+RUN export PYTHONPATH=/cache/models/torch_hub:$PYTHONPATH
+
+# install all requirements from repository is leading to reinstall openCV without CUDA support
+#RUN pip3 install -r https://raw.githubusercontent.com/hustvl/YOLOP/main/requirements.txt
+RUN pip3 install prefetch_generator matplotlib yacs
+
```
Comment (abramov-oleg) [inline samples/panoptic_driving_perception/docker/Dockerfile.l4t]: Extra line.

**Review by abramov-oleg — CHANGES_REQUESTED** [review]

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 7 above)
Comment (bwsw) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Why use fp32?

File: samples/panoptic_driving_perception/docker/Dockerfile.l4t (hunk 11 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/docker/Dockerfile.l4t]: Removed

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 10 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Removed

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 9 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Fixed

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 8 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Done

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 7 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Switched to half-precision

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 4 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Removed

File: samples/panoptic_driving_perception/pytorch_infer.py (hunk 3 above)
Comment (dorgun) [reply samples/panoptic_driving_perception/pytorch_infer.py]: Removed

**Review by bwsw — APPROVED** [review]
